client/src/transmitter/cast/webrtc.ts

```diff
@@ -0,0 +1,71 @@
+const log = (...data) => console.log('webrtc', ...data)
+
+async function iceHandshake(
+  streamKey: string,
+  localDesc: RTCSessionDescription
+) {
+  const qsKey = streamKey.indexOf('://') > 0 ? 'rtmp' : 'streamKey'
+  const answer = await fetch(
+    `https://webrtmp.justcast.it/webrtc/offer?${qsKey}=${streamKey}`,
+    {
+      method: 'POST',
+      body: JSON.stringify(localDesc),
+      headers: {
+        ['content-type']: 'application/json',
+      },
+    }
+  )
+  if (answer.status !== 200) {
+    throw new Error(`Error response from server: ${answer.status}`)
+  }
+  return (await answer.json()) as RTCSessionDescriptionInit
+}
+
+function castViaWebRTC(stream: MediaStream, streamKey: string): CastSession {
+  const pc = new RTCPeerConnection({
+    iceServers: [
+      {
+        urls: 'stun:stun.l.google.com:19302',
+      },
+    ],
+  })
+  const cast: CastSession = { stop: () => pc.close() }
+
+  stream.getTracks().forEach((track) => pc.addTrack(track, stream))
+
+  pc.oniceconnectionstatechange = () =>
+    log('ice connection state', pc.iceConnectionState)
+  pc.onicegatheringstatechange = () => {
+    if (pc.iceGatheringState !== 'complete') {
+      return
+    }
+    iceHandshake(streamKey, pc.localDescription)
+      .then(async (remoteDesc) => {
+        await pc.setRemoteDescription(new RTCSessionDescription(remoteDesc))
+      })
+      .catch(log)
+  }
+  pc.onconnectionstatechange = () => {
+    const state = pc.connectionState
+    log('connection state', state)
+    switch (state) {
+      case 'connected':
+        cast.onConnected?.call(cast)
+        break
+      case 'closed':
+        cast.onClosed?.call(cast)
+        break
+    }
+  }
+
+  pc.createOffer()
+    .then((offer) => pc.setLocalDescription(offer))
+    .catch((err) => {
+      log('create offer err', err)
+      cast.onClosed?.call(cast)
```

Comment:
Should we make sure it has connected before calling `onClosed`?

Reply:
I wanted to make sure to call `onClosed` so there is some way for the caller to know that the opening of the RTC had some problem connecting, so "onClosed" here would mean that there was some problem in the "connecting" state and that the connection is never going to happen. I didn't want to return a `Promise` from this function otherwise we wouldn't have a guarantee that the caller would be able to configure the `cast` object with their callbacks before any event is generated.

I think I'll rename these callback to `onStarted` and `onTerminated`, so that we remove this relation with some kind of connection that happens underneath, leave it as implementation detail, and also make the 2 callbacks a bit more independent, as being terminated before being able to start doesn't seem too weird to me. WDYT?